servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/DelegatingGrpcClientBuilder.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright © 2022 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.api;
+
+import java.time.Duration;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link GrpcClientBuilder} that delegates all methods to another {@link GrpcClientBuilder}.
+ *
+ * @param <U> the type of address before resolution (unresolved address)
+ * @param <R> the type of address after resolution (resolved address)
+ */
+public class DelegatingGrpcClientBuilder<U, R> implements GrpcClientBuilder<U, R> {
+
+    private GrpcClientBuilder<U, R> delegate;
+
+    public DelegatingGrpcClientBuilder(final GrpcClientBuilder<U, R> delegate) {
+        this.delegate = requireNonNull(delegate);
+    }
+
+    /**
+     * Returns the {@link GrpcClientBuilder} delegate.
+     *
+     * @return Delegate {@link GrpcClientBuilder}.
+     */
+    protected final GrpcClientBuilder<U, R> delegate() {
+        return delegate;
+    }
+
+    @Override
+    public String toString() {
+        return this.getClass().getSimpleName() + "{delegate=" + delegate() + '}';
+    }
+
+    @Override
+    public GrpcClientBuilder<U, R> initializeHttp(final HttpInitializer<U, R> initializer) {
+        delegate = delegate.initializeHttp(initializer);
+        return this;
+    }
+
+    @Override
+    public GrpcClientBuilder<U, R> defaultTimeout(final Duration defaultTimeout) {
+        delegate = delegate.defaultTimeout(defaultTimeout);
```

Comment:
When I had all builder methods as
```java
    @Override
    public GrpcClientBuilder<U, R> defaultTimeout(final Duration defaultTimeout) {
        delegate.defaultTimeout(defaultTimeout);
        return this;
    }
```
Spotbugs complained that the return value is discarded. I though to suppress it because all our builder implementations always return `this`. But then considered a use-case when users may have an immutable provider or builder impl that always return a new instance. LMK what you think. Should we support that and reassign the `delegate` or is it better to keep `delegate` as `final`?

Reply:
I don't see any alternative to assigning `delegate` with each delegate invocation. Eventually someone will get clever and implement a builder which doesn't return `this` and we will have to fix it anyway. I've had to fix very non-obvious problems more than once where failing to reassign the builder at the end of the chain meant that the operations in the chain were lost. It is always a frustrating bug to locate.